### PR TITLE
refactor(semver): generalize semver related structs

### PR DESCRIPTION
**Pull request by dsherret**
- Generalizes the npm version code (ex. `NpmVersion` -> `Version`, `NpmVersionReq` -> `VersionReq`). This is a slow refactor towards extracting out this code for deno specifiers and better usage in deno_graph.
- Removes `SpecifierVersionReq`. Consolidates `NpmVersionReq` and `SpecifierVersionReq` to just `VersionReq`
- Removes `NpmVersionMatcher`. This now just looks at `VersionReq`.
- Paves the way to allow us to create `NpmPackageReference`'s from a package.json's dependencies/dev dependencies (`VersionReq::parse_from_npm`).

### CODE REVIEW

File: cli/npm/resolution/reference.rs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+use deno_core::anyhow::Context;
+use deno_core::error::generic_error;
+use deno_core::error::AnyError;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::semver::VersionReq;
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct NpmPackageReference {
```
Comment (bartlomieju) [inline cli/npm/resolution/reference.rs:13]: Could you add a docstring here for posterity describing what a "reference" looks like?

File: cli/npm/resolution/reference.rs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    } else {
+      write!(f, "npm:{}", self.req)
+    }
+  }
+}
+
+#[derive(
+  Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize,
+)]
+pub struct NpmPackageReq {
```
Comment (bartlomieju) [inline cli/npm/resolution/reference.rs:108]: Ditto here, a docstring would be helpful for posterity to see a difference between a "reference" and a "req"

File: cli/npm/resolution/reference.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      Some(req) => write!(f, "{}@{}", self.name, req),
+      None => write!(f, "{}", self.name),
+    }
+  }
+}
+
+impl NpmPackageReq {
+  pub fn from_str(text: &str) -> Result<Self, AnyError> {
+    // probably should do something more targeted in the future
+    let reference = NpmPackageReference::from_str(&format!("npm:{text}"))?;
```
Comment (bartlomieju) [inline cli/npm/resolution/reference.rs:125]: I suggest we add a helper method like `from_str_without_prefix()` - there's a bunch of places (like lockfile) where we need to do these temporary allocations just to add `npm:` prefix which is stripped immediately after

**Review by bartlomieju — APPROVED** [review]
LGTM, very nice cleanup. Just a few requests for some more documentation

File: cli/npm/resolution/reference.rs (hunk 3 above)
Comment (dsherret) [reply cli/npm/resolution/reference.rs]: I extracted out and reused the common code.

**Pull request merged**